src/Hazelcast.Net/Serialization/Compact/CompactReader.cs

```diff
@@ -0,0 +1,577 @@
+        }
+
+        // reads an array of values of a class type that can be null (value is T[])
+        private T?[]? ReadArrayOfReference<T>(string name, FieldKind kind, Func<ObjectDataInput, T> read)
+        {
+            var field = GetValidField(name, kind);
+            var offset = _offsetReader!(_input, _offsetPosition, field.Index);
+            if (offset < 0) return null;
+
+            _input.MoveTo(DataStartPosition + offset);
+
+            var arrayDataLength = _input.ReadInt();
+            var arrayOffsetReader = GetOffsetReader(arrayDataLength);
+
+            var count = _input.ReadInt();
+            var items = new T?[count];
+
+            var arrayDataPosition = _input.Position;
+            var arrayOffsetPosition = arrayDataPosition + arrayDataLength;
+            for (var i = 0; i < count; i++)
+            {
+                var itemOffset = arrayOffsetReader(_input, arrayOffsetPosition, i);
+                if (itemOffset < 0)
+                {
+                    items[i] = default;
+                }
+                else
+                {
+                    _input.MoveTo(arrayDataPosition + itemOffset);
+                    items[i] = read(_input);
+                }
+            }
+
+            return items;
+        }
+
+        #endregion
+
+        private bool? ReadMaybeBoolean(string name)
+        {
+            var field = GetValidField(name);
+
+            if (field.Kind == FieldKind.Boolean)
+            {
+                var (position, offset) = (DataStartPosition + field.Offset, field.BitOffset);
+
+                // how we write:
+                //_output.MoveTo(position, BytesExtensions.SizeOfByte);
+                //var bits = (byte)(value ? 1 << offset : 0);
+                //var mask = (byte)(1 << offset);
+                //_output.WriteBits(bits, mask);
+
+                // how we read:
+                _input.MoveTo(position);
+                var bits = _input.ReadByte();
+                var mask = (byte)(1 << offset);
+                return (bits & mask) != 0;
+            }
+
+            if (field.Kind == FieldKind.NullableBoolean)
+            {
+                return ReadNullable(field, input => input.ReadBoolean());
+            }
+
+            throw new SerializationException($"Invalid kind \"{FieldKind.Boolean}\" or \"{FieldKind.NullableBoolean}\" for field \"{name}\" of schema {Schema}, which is of kind \"{field.Kind}\".");
+        }
+
+        public bool ReadBoolean(string name)
+        {
+            var value = ReadMaybeBoolean(name);
+            if (value.HasValue) return value.Value;
+            throw new SerializationException($"Null value for field \"{name}\" of schema {Schema}, which is of kind \"{FieldKind.NullableBoolean}\".");
+        }
+
+        public bool? ReadNullableBoolean(string name)
+            => ReadMaybeBoolean(name);
+
+        private T[]? ReadArrayOfBoolean<T>(SchemaField field, Func<bool, T> f)// where T : struct
```

Comment:
I'm confused with generic and `boolean` here. Is it expected the `T` is a type of collection or what?

Reply:
Oh, got it it's for nullable types.